File: core/parser.go
package core

import (
	"bytes"
	"errors"
	"regexp"
	"time"
)

type FileLanuages int

const (
	FileLanuagesEn FileLanuages = iota
	FileLanuagesZh
)

type clippingsParser struct {
	file           []byte
	lines          [][][]byte
	language       FileLanuages
	separator      []byte
	locationRegexp *regexp.Regexp
	chineseRegexp  *regexp.Regexp
}

func removeBOM(src []byte) []byte {
	return bytes.Trim(src, "\xef\xbb\xbf")
}

func NewClippingParser(src []byte) clippingsParser {
	return clippingsParser{
		file:          removeBOM(src),
		lines:         [][][]byte{},
		separator:     []byte("========"),
		chineseRegexp: regexp.MustCompile(`[\x{4E00}-\x{9FFF}|[\x{3000}-\x{303F}]`),
	}
}

func (c *clippingsParser) Prepare() error {
	lines := bytes.Split(c.file, []byte("\n"))
	temp := make([][]byte, 0)

	for _, line := range lines {
		line = bytes.TrimSpace(removeBOM(line))
		if bytes.Contains(line, c.separator) {
			c.lines = append(c.lines, temp)
			temp = [][]byte{}
		}
		if !bytes.Contains(line, c.separator) {
			temp = append(temp, line)
		}
	}

	if bytes.Contains(c.file, []byte("Your Highlight on")) {
		c.locationRegexp = regexp.MustCompile(`\d+(-?\d+)?`)
		c.language = FileLanuagesEn
	} else if bytes.Contains(c.file, []byte("您在")) {
		c.locationRegexp = regexp.MustCompile(`#?\d+(-?\d+)?`)
		c.language = FileLanuagesZh
	} else {
		return errors.New("哎呀呀，暂不支持非中英文的内容呢~")
	}

	return nil
}

func (c *clippingsParser) DoParse() (result []TClippingItem, err error) {
	for _, dataset := range c.lines {
		item := TClippingItem{}
		title := c.exactTitlte(dataset[0])
		item.Title = title

		pageAt, date, err := c.exactInfo(dataset[1])
		if err != nil {
			return result, err
		}
		item.PageAt = pageAt
		item.CreatedAt = date

		if len(dataset[3]) == 0 {
			continue
		}

		item.Content = string(dataset[3])
		result = append(result, item)
	}

	//   return this.result.filter(item => item.content && item.content !== "")
	return
}

func (c clippingsParser) exactTitlte(line []byte) string {
	STOP_WORDS := [][]byte{[]byte("("), []byte("（")}
	title := line

	for _, s := range STOP_WORDS {
		title = bytes.Split(title, s)[0]
	}

	title = bytes.TrimSpace(title)
	return string(title)
}

func (c clippingsParser) exactInfo(line []byte) (pageAt string, date time.Time, err error) {
	l := bytes.Split(line, []byte("|"))
	locationSection := l[0]
	dateSection := l[len(l)-1]
	locationResult := c.locationRegexp.FindStringSubmatch(string(locationSection))

	pageAt = locationResult[0]

	dateSection = bytes.Replace(dateSection, []byte("Added on "), []byte(""), 1)
	dateSection = bytes.Replace(dateSection, []byte("添加于 "), []byte(""), 1)

	timeLayout := KindleDateTimeENLayout

	if c.language == FileLanuagesZh {
		dateSection = []byte(c.chineseRegexp.ReplaceAllString(string(dateSection), "-"))
		dateSection = []byte(regexp.MustCompile(`-{2,10}`).ReplaceAllString(string(dateSection), ""))
		timeLayout = KindleDateTimeZHLayout
	}

	dateSection = bytes.TrimSpace(dateSection)

	date, err = time.Parse(timeLayout, string(dateSection))
	return
}

File: src/parser.rs
use chrono;
use regex::Regex;
use serde::{Deserialize, Serialize};
use std::error::Error;
use std::vec::Vec;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TClippingItem {
    pub title: String,
    pub content: String,
    #[serde(rename = "pageAt")]
    pub page_at: String,
    #[serde(rename = "createdAt", with = "datetime_parser_with_rfc3339")]
    pub created_at: chrono::NaiveDateTime,
}

mod datetime_parser_with_rfc3339 {
    use chrono::{DateTime, NaiveDateTime, TimeZone, Utc};
    use serde::{self, Deserialize, Deserializer, Serializer};
    pub fn serialize<S>(date: &NaiveDateTime, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let s = Utc.from_local_datetime(date).unwrap().to_rfc3339();
        serializer.serialize_str(&s)
    }
    pub fn deserialize<'de, D>(deserializer: D) -> Result<NaiveDateTime, D::Error>
    where
        D: Deserializer<'de>,
    {
        let s = String::deserialize(deserializer)?;
        let d = DateTime::parse_from_rfc3339(&s).unwrap();
        Ok(d.naive_utc())
    }
}

#[derive(Debug)]
enum KindleClippingLanguage {
    Zh,
    En,
}

struct ParserLanguageConfig {
    location: regex::Regex,
    language: KindleClippingLanguage,
}

pub fn do_parse(input: &str) -> Result<Vec<TClippingItem>, Box<dyn Error>> {
    let separator = "========";
    let lines: Vec<&str> = input.split('\n').collect();
    let mut grouped: Vec<Vec<String>> = Vec::new();
    let mut temp: Vec<String> = Vec::new();

    let la_config: ParserLanguageConfig = if input.contains("Your Highlight on") {
        let a = Regex::new(r"\d+(-?\d+)?").unwrap();
        ParserLanguageConfig {
            location: a,
            language: KindleClippingLanguage::En,
        }
    } else {
        let a = Regex::new(r"#?\d+(-?\d+)?").unwrap();
        ParserLanguageConfig {
            location: a,
            language: KindleClippingLanguage::Zh,
        }
    };

    // TODO: reduce O(n^2) to O(n)
    for l in lines {
        if l.contains(separator) {
            grouped.push(temp.clone());
            temp.clear();
        } else {
            temp.push(String::from(l));
        }
    }

    let mut result_list: Vec<TClippingItem> = vec![];
    let chinese_regex = Regex::new(r"[\x{4E00}-\x{9FFF}|\x{3000}-\x{303F}]").unwrap();

    let r = regex::Regex::new(r"\u{feff}").unwrap();
    for row in grouped {
        let content = row[3].trim();
        if content.is_empty() {
            continue;
        }

        let title = parse_title(&r.replace_all(&row[0], "").trim().to_string());
        let (location, dt) = parse_info(
            &row[1],
            &la_config.location,
            &la_config.language,
            &chinese_regex,
        )?;
        let item = TClippingItem {
            content: content.to_string(),
            title: title,
            page_at: location,
            created_at: dt,
        };
        result_list.push(item.clone());
    }

    Ok(result_list)
}

fn parse_title(line: &String) -> String {
    let stop_worlds: Vec<&str> = vec!["(", "（"];
    let mut title = line.clone();
    for w in stop_worlds {
        let nt: Vec<&str> = title.split(w).collect();
        title = nt.first().unwrap().to_string();
    }

    title.trim_end_matches(")").trim().to_string()
}

fn parse_info(
    line: &String,
    location_regex: &Regex,
    la: &KindleClippingLanguage,
    chinese_regex: &Regex,
) -> Result<(String, chrono::NaiveDateTime), Box<dyn Error>> {
    let ls: Vec<&str> = line.split("|").collect();
    let location_section = ls[0];

    let matched = location_regex.captures(location_section);
    if matched.is_none() {
        return Err("location not found".into());
    }
    let page_at_str = matched.unwrap().get(0).unwrap().as_str();

    let page_at = if page_at_str.starts_with("#") {
        page_at_str.to_string()
    } else {
        ("#".to_owned() + page_at_str).to_string()
    };

    let dt: chrono::NaiveDateTime;

    let date_section = ls[ls.len() - 1]
        .replace("Added on ", "")
        .replace("添加于 ", "");

    match la {
        KindleClippingLanguage::Zh => {
            let clocks_ext = {
                if date_section.contains("上午") {
                    "AM"
                } else {
                    "PM"
                }
            };
            let d = chinese_regex.replace_all(&date_section, "-");
            let mut f = Regex::new(r"-{2,10}")
                .unwrap()
                .replace_all(&d, "")
                .trim()
                .to_owned();
            f.push_str(" ");
            f.push_str(clocks_ext);
            // "2006-1-2 3:4:5"
            let parsed_dt = chrono::NaiveDateTime::parse_from_str(&f, "%Y-%m-%e %l:%M:%S %p")?;
            dt = parsed_dt;
        }
        KindleClippingLanguage::En => {
            let parsed_dt = chrono::NaiveDateTime::parse_from_str(
                &date_section.trim(),
                "%A, %B %e, %Y %l:%M:%S %p",
            )?;
            dt = parsed_dt;
        }
    }
    Ok((page_at.to_string(), dt))
}
